packages/web/src/pages/chat-page/components/ChatComposer.tsx

```diff
@@ -45,37 +56,104 @@ export const ChatSendButton = ({ disabled }: ChatSendButtonProps) => {
       type='submit'
       size='m'
       icon={IconSend}
-      color='staticWhite'
-      iconCss={{ position: 'relative', left: -1 }}
+      color='active'
     />
   )
 }
 
+const ComposerText = ({
+  color,
+  children
+}: Pick<TextProps, 'color' | 'children'>) => {
+  if (children === '\n') {
+    return <br />
+  }
+  return (
+    <Text style={{ whiteSpace: 'pre-wrap' }} color={color}>
+      {children}
+    </Text>
+  )
+}
+
 export const ChatComposer = (props: ChatComposerProps) => {
   const { chatId, presetMessage, onMessageSent } = props
   const dispatch = useDispatch()
+
   const [value, setValue] = useState(presetMessage ?? '')
+  const [focused, setFocused] = useState(false)
+
+  // Maintain bidirectional maps of audius links to human readable format
+  const linkToHuman = useRef<{ [key: string]: string }>({}).current
+  const humanToTrack = useRef<{
+    [key: string]: { link: string; track: Track }
+  }>({}).current
+
+  // The track id used to render the composer preview
+  const [trackId, setTrackId] = useState<ID | null>(null)
+
   const ref = useRef<HTMLTextAreaElement>(null)
   const chatIdRef = useRef(chatId)
 
   const handleChange = useCallback(
-    (e: ChangeEvent<HTMLTextAreaElement>) => {
-      setValue(e.target.value)
+    async (e: ChangeEvent<HTMLTextAreaElement>) => {
+      const originalValue = e.target.value
+      setValue(originalValue)
+
+      const { matches } = matchAudiusLinks({
```

Comment:
linkify only takes care of the hyperlinking now, so nothing to do with the jitter actually. that comes from our custom unfurler code. looking into this change to see how bad it'll be

Reply:
this is challenging!
here is a WIP: https://github.com/AudiusProject/audius-protocol/compare/rj-pay-3227...rj-pay-3227-linkify?expand=1

the biggest flaw is that linkify doesn't do a good job for things that are not actually links and don't match _their_ regex stuff. Since the underlying `textarea` needs to get the link stripped out so that cursors / highlighting match the overlayed display element, we need to mod linkify to detect anything in our "humanToLink" mapping. There is a plugin interface, but it would require a bigger lift https://linkify.js.org/docs/linkifyjs.html#linkifyregisterplugin-name-plugin. At that point it maybe doesn't end up benefitting us too much.

This would also make doing `@`-mentions for comments harder I think as well. Down to keep plucking at it, but I think this custom impl is solid enough for now

Reply:
Yeah I think linkify works better for link => name but not the reverse

If it helps, it might be easier to use the non-react version of linkify to do the custom things

Reply:
could be cool to abstract this into a hook, with an interface like:

`const { formatEntityLinks, restoreEntityLinks, entityLinks } = useEntityLinkFormatter()`

Where `formatEntityLinks(value)` takes in a value, parses out entity (eg track) links, replaces them with titles, `restoreEntityLinks` puts the links back in place of the titles, and `entities` is the map of link => entity found after `formatEntityLinks` runs